tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -407,6 +408,303 @@ impl ServerKey {
+                    // to have less noise growth later
+                    (propagation_result as u64) << bits_of_message
+                });
+        self.key.unchecked_apply_lookup_table_bivariate(
+            last_lhs_block,
+            last_rhs_block,
+            &last_block_inner_propagation_lut,
+        )
+    }
+
+    // - last_block_inner_propagation must be the result of generate_last_block_inner_propagation
+    // - last_block_input_carry: carry that the last pair of blocks (lhs, rhs) receives as input
+    // - last_block_output_carry: carry that the last pair of blocks (lhs, rhs) output
+    //
+    // Returns whether the subtraction overflowed
+    //
+    // See [unchecked_signed_overflowing_sub_parallelized] for more context
+    fn resolve_signed_overflow(
+        &self,
+        mut last_block_inner_propagation: Ciphertext,
+        last_block_input_carry: &Ciphertext,
+        last_block_output_carry: &Ciphertext,
+    ) -> Ciphertext {
+        let bits_of_message = self.key.message_modulus.0.ilog2();
+
+        let resolve_overflow_lut = self.key.generate_lookup_table(|x| {
+            let carry_propagation = x >> bits_of_message;
+            let output_carry_of_block = (x >> 1) & 1;
+            let input_carry_of_block = x & 1;
+
+            // Resolve the carry that the last bit actually receives as input
+            let input_carry_to_last_bit = if carry_propagation == OutputCarry::Propagated as u64 {
+                input_carry_of_block
+            } else if carry_propagation == OutputCarry::Generated as u64 {
+                1
+            } else {
+                0
+            };
+
+            u64::from(input_carry_to_last_bit != output_carry_of_block)
+        });
+
+        let x = self.key.unchecked_scalar_mul(last_block_output_carry, 2);
+        self.key
+            .unchecked_add_assign(&mut last_block_inner_propagation, &x);
+        self.key
+            .unchecked_add_assign(&mut last_block_inner_propagation, last_block_input_carry);
+        self.key
+            .apply_lookup_table(&last_block_inner_propagation, &resolve_overflow_lut)
+    }
+
+    // This is the implementation of overflowing sub when we can use parallel carry
+    // propagation.
+    //
+    // It is in its own function so that it can be tested, as the main entry point
+    // unchecked_signed_overflowing_sub may select non parallel version if lhs
+    // does not have enough block.
+    pub(crate) fn unchecked_signed_overflowing_sub_parallelized_impl(
+        &self,
+        lhs: &SignedRadixCiphertext,
+        rhs: &SignedRadixCiphertext,
+    ) -> (SignedRadixCiphertext, Ciphertext) {
+        // This assert is here because this overflow computation requires these preconditions
+        // which is_eligible_for_parallel_single_carry_propagation, but it could change in the
+        // future
+        assert!(self.key.message_modulus.0 >= 4 && self.key.carry_modulus.0 >= 4);
```

Comment:
No because is_eligible_for_parallel_single_carry_propagation  checks for other stuff as well (like number of blocks to select between fully parallelized propagation or sequential)

Reply:
should this sub check be a function that is called both here and in the is_elligible for parallel etc ?

Not saying we need it but in case some pre conditions change ?

Reply:
`assert!(self.key.message_modulus.0 >= 4 && self.key.carry_modulus.0 >= 4);` is a precondition for both parallel propagation and `unchecked_signed_overflowing_sub_parallelized_impl`

If the precondition changes for parallel_propagation (and is updated in is_eligible_for_parallel_single_carry_propagation), that does not mean the precondition for `unchecked_signed_overflowing_sub_parallelized_impl` also changes

Reply:
ok then, I was guessing something like this may happen, all good